torch/csrc/jit/codegen/cuda/manager.cpp

```diff
@@ -264,7 +305,17 @@ void runCudaFusionGroup(const Node* fusion_node, Stack& stack) {
 
   if (useFallback()) {
     try {
-      run_fusion();
+      // if fusion failed once, it's likely to fail again; and failures are
+      // slow. So if the fusion fails, then record the failure and always use
+      // the fallback instead
+      int32_t kernel_id = fusion_node->i(attr::cache_id);
```

Comment:
So this one will trigger fallback, which is functionally fine since it's caught. But it does not go through the cached fallback and it'll be slow.

Reply:
I think we can change something here:
https://github.com/pytorch/pytorch/blob/9bf1497d05045be03e5281de4152667b15505d1b/torch/csrc/jit/codegen/cuda/manager.cpp#L236
```
CudaFusionManager::getManager().unregisterCacheId(graph);
```
vvv
```
if (!fusion_node->hasAttribute(attr::cache_id)) {
  int32_t fusion_cache_id =
      CudaFusionManager::getManager().registerOrGetCacheId(graph);
  fusion_node->i_(attr::cache_id, fusion_cache_id);
}
CudaFusionManager::getManager().unregisterCacheId(graph);
```

And then we can assert at the end of the function that `fusion_node` should always have the attribute.

This is a little bit unfortunate that we are not re-using the fallback code for identical graph, but we can leave that as a TODO for later.